### DISCUSSION TITLE
Specify rainbowkit modals parent element.

### DISCUSSION THREAD

**Discussion by Linch1** [rainbow-me/rainbowkit#1790] [Ideas]
Hello and thanks for this great tool!

I am currently working on integrating RainbowKit into a component that resides within a Shadow DOM. I've encountered an issue where, upon triggering a RainbowKit modal, it is appended directly to the body of the DOM instead of within the Shadow DOM where the component is located. This results in the modal missing the necessary styles, as these are defined within the Shadow DOM and are not accessible to elements outside of it.

I have provided screenshots below to illustrate the problem. The first screenshot shows the modal rendered without the appropriate styles, and the second screenshot demonstrates that the theme styles, which I believe are responsible for the modal's appearance, are encapsulated within the Shadow DOM and thus not applied to the modal placed in the DOM body.

![Screenshot 2024-02-22 014443](https://github.com/rainbow-me/rainbowkit/assets/47044850/d154d0fe-d85d-408d-885a-5574fcefa2ea)
![Screenshot 2024-02-22 014755](https://github.com/rainbow-me/rainbowkit/assets/47044850/f68cc504-edd4-4c4c-959a-3f5496811b62)

If i manually move the theme styles in the body, so outside of the shadow dom, the rainbowkit modal is rendered correctly.

Could you advise on how to specify a parent element for RainbowKit modals? 
This would allow me to render them inside the Shadow DOM, ensuring the modal has access to the required styles.

**Comment by 0xmkh** [comment]
@Linch1 Thanks for the kind words! I'm happy to help you :)

I see what you're trying to achieve here. Unfortunately the RainbowKit needs to access the elements outside of it such as the `<style>` elements you have in the 2nd picture since we use vanilla extract library to generate those styles.

I think what you could try to do is to append your own theme `<style />` which is shown in the docs [here](https://www.rainbowkit.com/docs/custom-theme#creating-custom-theme-selectors) and essentially append that into your shadow dom ?

**Reply by 0xmkh** [reply]
Something like this:

```
  <div data-rk="">
    <style
      dangerouslySetInnerHTML={{
        __html: [`[data-rk]{${cssStringFromTheme(lightTheme)}}`].join(""),
      }}
    />
    <RainbowKitProvider theme={null}>
      <App />
    </RainbowKitProvider>
  </div>
```

**Reply by Linch1** [reply]
Thank you for answer! 

Unfortunately, any style I add within the Shadow DOM remains invisible to the tags outside in the main document body. This means that if a modal is situated in the body and the style is confined to the Shadow DOM, they don't interact. So i have to place it outside of the shadow dom to make it visible to the modal.

Currently, I'm duplicating the style tag from the Shadow DOM into the body as soon as it is added to the dom, and everything appears to work correctly, but it feels like a workaround. Hence, I'm was curious if there's a way to designate where these styles should apply. 

Perhaps this could be a feature worth considering for future development :)

**Reply by 0xmkh** [reply]
Mhmm i see. Yeah that won't work unfortunately. Is this something you would like us to have a look at so you can apply the styles to your shadow DOM use case ?

Like you said you can just duplicate for now which is just a workaround, but we will have a look into this 🙏

**Reply by DJHellscream** [reply]
Has there been any progress with this? I am also trying to use rainbowkit inside of a webcomponent and can't get the default styling to apply.

**Reply by 0xmkh** [reply]
We haven't done anything yet no. We'll get back to you if we're making any progress on this.

**Reply by DJHellscream** [reply]
Okay, thank you for the update!

**Comment by tiendn** [comment]
Hi there,
I'm having an issue with customizing the AccountModal in Rainbow, such as adding a button or an icon.
Currently, it seems impossible to make these changes directly from React root. 
Do you have any suggestions for how to approach this?

**Reply by tiendn** [reply]
ah, I found a solution. Use eventListener for trigger click btn, then wait for Modal insert into DOM. Then find it and style or do st.
Very nice, tks Rainbow Team!
```js
 useEffect(() => {
        const accountButton = document.querySelector('[data-testid="rk-account-button"]');
        if (accountButton) {
            accountButton.addEventListener('click', () => {
                console.log('clicked');
                setTimeout(() => {
                    const div = document.querySelector(
                        'body > div:nth-child(9) > div > div > div:nth-child(2) > div > div > div > div > div:nth-child(3)',
                    );
                    console.log(div);
                }, 3000);
            });
        }
    }, [account]);
 ```

**Comment by Hamad-FE** [comment]
Hello,

i am using rainbow kit in a widget that is implemented to be embedded in other websites. It uses shadow dom. I am facing the issue that rainbow kit styles are off when i use it in my widget. how can i fix it any solution for it??
